Links: do not reposition container on selection change

ellatrix opened the pull request:
## Description

Fixes #14872. It also makes sure that the container is positioned relative to the link element rather than the selection/caret, so it doesn't move when you move selection inside a link.

There are several steps that came to the fix.

1. I wanted to modify `PositionedAtSelection` so that it accepts a `selector` prop, to select the closest matching element to calculate the position from rather than the caret.
2. In order to not recreate `PositionedAtSelection` for every change in selection, we need to use the format object reference as an indicator to recreate, as each DOM element is a different object instance.
3. Unfortunately, React does not accept object references as keys, so we have to derive a simpler state to use as the key. I used `getDerivedStateFromProps` to bump the key every time the active format object changes.
4. Something was wrong in setting the `activeFormats` in `onSelectionChange`. The formats need to be taken from the current value to maintain references, not from the created value (from DOM) on the selection change.
5. To ensure object references persist in the RichText component across selection changes and edits, I've taken out recreating the rich text value for every change. This actually results in other positive changes: no more `formatToValue` call for every input event, and no more need memoize it, and no more need for `updateFormats`, introduced in #14411.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

ellatrix committed:
Links: do not reposition container on selection change

ellatrix committed:
Use caret position when adding link

ellatrix commented:
The only remaining issue here is that the toolbar rerenders when you typing in the element. Fixing that now.

ellatrix committed:
WIP

aduth commented on packages/format-library/src/link/inline.js:
```diff
@@ -222,7 +236,12 @@ class InlineLinkUI extends Component {
 
 		return (
 			<PositionedAtSelection
-				key={ `${ value.start }${ value.end }` /* Used to force rerender on selection change */ }
```
Possible dumb question: Couldn't a minimal solution just be to remove `key`?

ellatrix replied:
What if you move the caret from one link to another?

aduth replied:
> What if you move the caret from one link to another?

That's true. I wonder then if it could either act more like the `Autocomplete` component in [rendering `Popover` with a `getAnchorRect`](https://github.com/WordPress/gutenberg/blob/0b93ad47797a1ba2cea3c83a7f973d33abcd351f/packages/components/src/autocomplete/index.js#L469) that updates with the selection (but keeps the same element so not to unmount and re-trigger animation), or if `PositionedAtSelection` should update its styling in response to selection changes rather than [once at mount](https://github.com/WordPress/gutenberg/blob/0b93ad47797a1ba2cea3c83a7f973d33abcd351f/packages/components/src/positioned-at-selection/index.js#L50-L52).

I'm not totally sure the ramifications of either of these alternatives. Just trying to consider in terms of the least potential for impact.

ellatrix committed:
Fix annotations

aduth commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -459,12 +470,18 @@ export class RichText extends Component {
+				end,
+			};
+			delete newValue.activeFormats;
+			const activeFormats = getActiveFormats( newValue );
```
If we're working with a shallow clone, seems we might as well assign `activeFormats` directly, rather than create another shallow clone below?

```js
const newValue = {
	...value,
	activeFormats: getActiveFormats( newValue ),
	start,
	end,
};
```

aduth commented on packages/block-editor/src/components/rich-text/index.js:
```diff
@@ -763,6 +781,7 @@ export class RichText extends Component {
 		// In all other cases, prevent default behaviour.
 		event.preventDefault();
 
+		const formats = createPrepareEditableTree( this.props )( value );
```
I don't understand why this change is here.

ellatrix replied:
I'll comment, it's to make sure the boundaries work for annotations.

ellatrix committed:
Fix initial selection

ellatrix commented:
Updated the description:

> To ensure object references persist in the RichText component across selection changes and edits, I've taken out recreating the rich text value for every change. This actually results in other positive changes: no more formatToValue call for every input event, and no more need memoize it, and no more need for updateFormats, introduced in #14411.

ellatrix commented:
While I think we eventually want all the changes here to RichText, I'm going to make an alternative PR that doesn't touch as much existing code as it does now.

aduth commented:
Needs to be rebased after #14938

aduth commented:
(I'm assuming this doesn't _need_ to land for Gutenberg 5.5 and have removed it from the milestone accordingly)

ellatrix commented:
Closing for now.
